dll/win32/shell32/wine/control.c

```diff
@@ -736,6 +737,96 @@ static	void	Control_DoWindow(CPanel* panel, HWND hWnd, HINSTANCE hInst)
 }
 #endif
 
+#ifdef __REACTOS__
+
+/** Structure for in and out data when
+ *  search for the cpl dialog of first instance
+ */
+typedef struct tagAppDlgFindData
+{
+    PCWSTR  szAppFile;  /**< Full path to applet library as second search value */
+    HWND    hDlgResult; /**< Returned dialog handle or NULL if not found */
+} AppDlgFindData;
+
+/**
+ * This function searches for an applet library and
+ * compares the file paths.
+ * @param hwnd The handle of the applet dialog
+ * @param szAppFile Full path of applet library
+ * @return TRUE if correct applet library otherwise FALSE
+ */
+static BOOL
+Control_FindAndCheckAppFile(
+    _In_ HWND   hwnd,
+    _In_ PCWSTR szAppFile
+)
+{
+    DWORD dwProcessId;
+    GetWindowThreadProcessId(hwnd, &dwProcessId);
+    if (dwProcessId)
+    {
+        HANDLE hProcess;
+        hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, dwProcessId);
+        if (hProcess)
+        {
+            HMODULE hMods[1024];
+            WCHAR szModName[MAX_PATH];
+            DWORD cbNeeded;
+
+            if (EnumProcessModules(hProcess, hMods, sizeof(hMods), &cbNeeded))
+            {
+
+                for (unsigned int i = 0; i < (cbNeeded / sizeof(HMODULE)); i++)
+                {
+                    // Get the full path to the module's file.
+                    if (GetModuleFileNameExW(hProcess, hMods[i], szModName, _countof(szModName)))
+                    {
+                        if (_wcsicmp(szModName, szAppFile) == 0)
+                        {
+                            return TRUE;
+                        }
+                    }
+                }
+            }
+            CloseHandle(hProcess);
+        }
+    }
+    return FALSE;
+}
+
+
+/**
+ * Callback function to search applet dialog
+ * @param hwnd A handle to a top-level window.
+ * @param lParam Pointer of AppDlgFindData
+ * @return TRUE: continue enumeration, FALSE: stop enumeration
+ */
+static BOOL CALLBACK
+Control_EnumWinProc(
+    _In_ HWND   hwnd,
+    _In_ LPARAM lParam
+)
+{
+    WCHAR szClassName[256] = L"";
+    if (GetClassNameW(hwnd, szClassName, _countof(szClassName)))
+    {
+        // check of first search value
+        if (wcscmp(L"#32770", szClassName) == 0)
+        {
+            AppDlgFindData* pData = (AppDlgFindData*)lParam;
+            // check of second search value
+            if (Control_FindAndCheckAppFile(hwnd, pData->szAppFile))
```

Comment:
I don't think this is a good way to determine if a control panel applet is already running,
because now you are opening **each process** that spawns a dialog, to check if it has a certain dll loaded.

And a small nitpick:
A comment like `// check of second search value` does not really add value to the code, either explain _why_ something is done, but just explaining what can be read in the next line does not add much value.

Reply:
Control_FindAndCheckAppFile is called only for processes whose dialog class is '#32770'. So **not all processes** will be opened.
However, I can move the class check from Control_EnumWinProc to Control_FindAndCheckAppFile to eliminate this problem, which can occur when Control_FindAndCheckAppFile is called from another location.

Reply:
> Control_FindAndCheckAppFile is called only for processes whose dialog class is '#32770'. So **not all processes** will be opened. However, I can move the class check from Control_EnumWinProc to Control_FindAndCheckAppFile to eliminate this problem, which can occur when Control_FindAndCheckAppFile is called from another location.

In practice this means any application that opens a dialog.

To me it does not make sense to open every process, and look for loaded modules for something like this.
I would assume that windows attaches a window property to a dialog or something like that,
which can be retrieved by all applications without having to open the process.

Reply:
It may be more probable that the Windows shell maintains a list of the control panels it opened, and when a "new" one gets opened, it is checked against that maintained list, and if present, the existing CPL dialog gets focused instead.

Reply:
How many dialogs of class '#32770' will be open at the same time in practice? I guess between 1 and 5. Does browsing 5 processes bring the entire operating system to its knees?
Dialog class '#32770' is a system class used by operating system applications. Normal programs (Notepad, Wordpad, Paint, ...) use their own classes for their dialogs, except for those coming from comdlg32.dll (FileOpen, FileSave, Print, ...).

My examination of the CPL dialogs concerning the attributes GWL_STYLE and GWL_EXSTYLE did not reveal any peculiarities compared to the other dialogs of the class '#32770'.

The GWL_USERDATA element is too unspecific and may already be used for the tasks of the applet itself.

How do we want to proceed now?
a) Discard the code and ignore the problem of individual instances and the possible problems of parallel access to configuration data.
b) Return to the initial implementation with the multi-language problem.
c) Take over after moving the class check to Control_FindAndCheckAppFile.
d) Unchanged takeover.

No decision is a bad decision.

Reply:
> It may be more probable that the Windows shell maintains a list of the control panels it opened, and when a "new" one gets opened, it is checked against that maintained list, and if present, the existing CPL dialog gets focused instead.

A typical chicken - egg problem.

Only after the Windows shell has given control to the CPL applet, this applet itself creates the dialog. When the dialog is closed, control returns to the Windows shell for cleanup.

How should the applet pass its dialog handle to the Windows shell so that it can keep a list of open dialogs.

--- Code snipped ---
            if (!applet->proc(applet->hWnd, CPL_STARTWPARMSW, sp, (LPARAM)extraPmts))
                applet->proc(applet->hWnd, CPL_DBLCLK, sp, applet->info[sp].data);
---
The applet call is blocking.

Reply:
![image](https://user-images.githubusercontent.com/610685/160893839-772f31e9-f759-47e4-8ade-e7c7ead973a9.png)

These atoms appear on Windows Server 2003 SP2 when opening the first .cpl file,
so they will most likely be used with `SetProp` and `GetProp`:
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setpropW
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getpropw

Reply:
> How many dialogs of class '#32770' will be open at the same time in practice? I guess between 1 and 5. Does browsing 5 processes bring the entire operating system to its knees? Dialog class '#32770' is a system class used by operating system applications. 

'#32770' is a system class, but not in the way that it is only in use by the operating system.
It is used for ANY window that is created as a Dialog instead of as a Window (So not using `CreateWindow`, but any of the `DialogBoxXXX` functions



And no, this does not 'bring the operating system to it's knees', which you would have seen when testing,
but opening a process like this is simply not always possible.
What if the process is elevated / has different permissions?

Reply:
> These atoms appear on Windows Server 2003 SP2 when opening the first .cpl file, so they will most likely be used with `SetProp` and `GetProp`: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setpropW https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getpropw

The functions `SetProp` and `GetProp` are new to me. The only method I know to piggyback data to a window is based on GWL_USERDATA.

I will need a little time to get to grips with the functions. Maybe there is indeed a more elegant solution.

Reply:
Apimonitor might be of use here, if you add window related functions (SetProp, GetProp, FindWindow, EnumWindows, GetWindowClassname, ....) and all atom related functions, you can probably deduce what / how windows this is doing.

Reply:
In addition to what Mark pointed (the SetProp/GetProp stuff), one should also consider the window argument passed to the `Control_RunDll` call (which corresponds to the main window of the controlling application for the CPL).

Reply:
> It may be more probable that the Windows shell maintains a list of the control panels it opened, and when a "new" one gets opened, it is checked against that maintained list, and if present, the existing CPL dialog gets focused instead.

Unless it could be proven that this would be exactly what Windows does, I do think a stateless solution is preferable.